File: Pinball App 2/Pinball App 2/gameroom/GameRoomVariantPresentationSupport.swift
import SwiftUI

struct GameRoomVariantPill: View {
    enum Style {
        case mini
        case standard
        case machineTitle
        case editSelector

        var sharedStyle: AppVariantPillStyle {
            switch self {
            case .mini:
                return .mini
            case .standard:
                return .standard
            case .machineTitle:
                return .machineTitle
            case .editSelector:
                return .editSelector
            }
        }
    }

    let label: String
    var style: Style = .standard

    var body: some View {
        if style == .mini || style == .standard {
            Text(compactLabel)
                .font(style.sharedStyle.font)
                .foregroundStyle(.white)
                .lineLimit(1)
                .truncationMode(.tail)
                .padding(.horizontal, style.sharedStyle.horizontalPadding)
                .padding(.vertical, style.sharedStyle.verticalPadding)
                .frame(maxWidth: style == .mini ? nil : 84)
                .background(AppTheme.brandGold.opacity(0.20), in: Capsule())
                .overlay(
                    Capsule()
                        .stroke(AppTheme.brandGold.opacity(0.42), lineWidth: 0.8)
                )
                .offset(y: style.sharedStyle.verticalOffset)
        } else {
            AppVariantPill(
                title: compactLabel,
                style: style.sharedStyle,
                maxWidth: 84
            )
        }
    }

    private var compactLabel: String {
        let trimmed = label.trimmingCharacters(in: .whitespacesAndNewlines)
        let maxAllowed = 7
        guard trimmed.count > maxAllowed else { return trimmed }
        let prefix = String(trimmed.prefix(max(0, maxAllowed - 1)))
        return prefix + "…"
    }
}

func gameRoomVariantBadgeLabel(variant: String?, title: String) -> String? {
    if let variant {
        let cleanedVariant = variant.trimmingCharacters(in: .whitespacesAndNewlines)
        if !cleanedVariant.isEmpty,
           cleanedVariant.lowercased() != "null",
           cleanedVariant.lowercased() != "none",
           cleanedVariant.lowercased() != "premium/le",
           cleanedVariant.lowercased() != "premium le",
           cleanedVariant.lowercased() != "premium-le" {
            return cleanedVariant
        }
    }

    let loweredVariant = variant?.lowercased() ?? ""
    let loweredTitle = title.lowercased()
    let source = "\(loweredVariant) \(loweredTitle)"

    if source.contains("limited edition") ||
        source.contains("(le") ||
        source.hasSuffix(" le") ||
        source.contains(" le)") {
        return "LE"
    }
    if source.contains("premium") {
        return "Premium"
    }
    if source.contains("(pro") ||
        source.hasSuffix(" pro") ||
        source.contains(" pro)") ||
        loweredVariant == "pro" {
        return "Pro"
    }
    return nil
}

func gameRoomVariantBadgeLabel(for machine: OwnedMachine) -> String? {
    gameRoomVariantBadgeLabel(variant: machine.displayVariant, title: machine.displayTitle)
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/gameroom/GameRoomVariantPresentationSupport.kt
package com.pillyliu.pinprofandroid.gameroom

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.pillyliu.pinprofandroid.ui.AppVariantPill
import com.pillyliu.pinprofandroid.ui.AppVariantPillStyle
import com.pillyliu.pinprofandroid.ui.DropdownOption
import com.pillyliu.pinprofandroid.ui.DropdownOptionGroup
import com.pillyliu.pinprofandroid.ui.GroupedAnchoredDropdownFilter
import com.pillyliu.pinprofandroid.ui.PinballThemeTokens

internal enum class VariantPillStyle {
    Mini,
    Standard,
    MachineTitle,
    EditSelector,
}

@Composable
internal fun ManufacturerFilterDropdown(
    selectedText: String,
    modernOptions: List<GameRoomCatalogManufacturerOption>,
    classicPopularOptions: List<GameRoomCatalogManufacturerOption>,
    otherOptions: List<GameRoomCatalogManufacturerOption>,
    onSelect: (String?) -> Unit,
    modifier: Modifier = Modifier,
) {
    val groups = buildList {
        add(DropdownOptionGroup(options = listOf(DropdownOption("", "All Manufacturers"))))
        if (modernOptions.isNotEmpty()) {
            add(DropdownOptionGroup("Modern", modernOptions.map { DropdownOption(it.id, it.name) }))
        }
        if (classicPopularOptions.isNotEmpty()) {
            add(DropdownOptionGroup("Classic Popular", classicPopularOptions.map { DropdownOption(it.id, it.name) }))
        }
        if (otherOptions.isNotEmpty()) {
            add(DropdownOptionGroup("Other", otherOptions.map { DropdownOption(it.id, it.name) }))
        }
    }
    GroupedAnchoredDropdownFilter(
        selectedText = selectedText,
        groups = groups,
        onSelect = { selection -> onSelect(selection.ifEmpty { null }) },
        modifier = modifier,
    )
}

@Composable
internal fun VariantPillDropdown(
    selectedLabel: String,
    options: List<String>,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    var expanded by remember { mutableStateOf(false) }
    Box(modifier = modifier) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .clickable { expanded = true },
            contentAlignment = Alignment.CenterEnd,
        ) {
            GameRoomVariantPill(label = selectedLabel, style = VariantPillStyle.EditSelector)
        }
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            options.forEach { option ->
                DropdownMenuItem(
                    text = { Text(option, maxLines = 1, overflow = TextOverflow.Ellipsis) },
                    onClick = { onSelect(option); expanded = false },
                )
            }
        }
    }
}

@Composable
internal fun GameRoomVariantPill(
    label: String,
    style: VariantPillStyle,
    modifier: Modifier = Modifier,
) {
    val compactLabel = compactVariantLabel(label)
    val colors = PinballThemeTokens.colors
    if (style == VariantPillStyle.Mini || style == VariantPillStyle.Standard) {
        Text(
            text = compactLabel,
            color = Color.White,
            style = if (style == VariantPillStyle.Mini) {
                MaterialTheme.typography.labelSmall.copy(fontSize = 10.sp)
            } else {
                MaterialTheme.typography.labelSmall
            },
            fontWeight = FontWeight.SemiBold,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = modifier
                .then(if (style == VariantPillStyle.Mini) Modifier else Modifier.widthIn(max = 84.dp))
                .background(
                    colors.brandGold.copy(alpha = 0.20f),
                    RoundedCornerShape(999.dp),
                )
                .border(1.dp, colors.brandGold.copy(alpha = 0.42f), RoundedCornerShape(999.dp))
                .padding(horizontal = if (style == VariantPillStyle.Mini) 6.dp else 8.dp, vertical = 3.dp)
                .offset(y = (-1).dp),
        )
        return
    }

    AppVariantPill(
        label = compactLabel,
        style = when (style) {
            VariantPillStyle.Mini -> AppVariantPillStyle.Mini
            VariantPillStyle.Standard -> AppVariantPillStyle.Standard
            VariantPillStyle.MachineTitle -> AppVariantPillStyle.MachineTitle
            VariantPillStyle.EditSelector -> AppVariantPillStyle.EditSelector
        },
        modifier = modifier,
        maxWidth = 84.dp,
    )
}

internal fun gameRoomVariantBadgeLabel(variant: String?, title: String): String? {
    val explicit = variant?.trim().orEmpty()
    if (explicit.isNotBlank() &&
        !explicit.equals("null", ignoreCase = true) &&
        !explicit.equals("none", ignoreCase = true) &&
        !explicit.equals("premium/le", ignoreCase = true) &&
        !explicit.equals("premium le", ignoreCase = true) &&
        !explicit.equals("premium-le", ignoreCase = true)
    ) {
        return explicit
    }

    val source = "${variant.orEmpty().lowercase()} ${title.lowercase()}"
    return when {
        source.contains("limited edition") || source.contains("(le") || source.endsWith(" le") || source.contains(" le)") -> "LE"
        source.contains("premium") -> "Premium"
        source.contains("(pro") || source.endsWith(" pro") || source.contains(" pro)") || variant.equals("pro", ignoreCase = true) -> "Pro"
        else -> null
    }
}

private fun compactVariantLabel(label: String): String {
    val trimmed = label.trim()
    val maxAllowed = 7
    if (trimmed.length <= maxAllowed) return trimmed
    return "${trimmed.take((maxAllowed - 1).coerceAtLeast(0))}…"
}
